source/simulator/melt.cc

```diff
@@ -0,0 +1,1090 @@
+                                          internal::Assembly::Scratch::AdvectionSystem<dim>  &scratch,
+                                          internal::Assembly::CopyData::AdvectionSystem<dim> &data) const
+    {
+      const Introspection<dim> &introspection = this->introspection();
+      const bool use_bdf2_scheme = (this->get_timestep_number() > 1);
+      const unsigned int n_q_points = scratch.finite_element_values.n_quadrature_points;
+      const unsigned int advection_dofs_per_cell = data.local_dof_indices.size();
+
+      const double time_step = this->get_timestep();
+      const double old_time_step = this->get_old_timestep();
+
+      const unsigned int solution_component
+        = (advection_field.is_temperature()
+           ?
+           introspection.component_indices.temperature
+           :
+           introspection.component_indices.compositional_fields[advection_field.compositional_variable]
+          );
+
+      const FEValuesExtractors::Scalar solution_field
+        = (advection_field.is_temperature()
+           ?
+           introspection.extractors.temperature
+           :
+           introspection.extractors.compositional_fields[advection_field.compositional_variable]
+          );
+
+      MaterialModel::MeltOutputs<dim> *melt_outputs = scratch.material_model_outputs.template get_additional_output<MaterialModel::MeltOutputs<dim> >();
+
+      std::vector<Tensor<1,dim> > fluid_velocity_values(n_q_points);
+      FEValuesExtractors::Vector ex_u_f = introspection.variable("fluid velocity").extractor_vector();
+      scratch.finite_element_values[ex_u_f].get_function_values (this->get_solution(),fluid_velocity_values);
+
+      for (unsigned int q=0; q<n_q_points; ++q)
+        {
+          // precompute the values of shape functions and their gradients.
+          // We only need to look up values of shape functions if they
+          // belong to 'our' component. They are zero otherwise anyway.
+          // Note that we later only look at the values that we do set here.
+          for (unsigned int k=0; k<advection_dofs_per_cell; ++k)
+            {
+              scratch.grad_phi_field[k] = scratch.finite_element_values[solution_field].gradient (scratch.finite_element_values.get_fe().component_to_system_index(solution_component, k),q);
+              scratch.phi_field[k]      = scratch.finite_element_values[solution_field].value (scratch.finite_element_values.get_fe().component_to_system_index(solution_component, k), q);
+            }
+
+          double bulk_density = scratch.material_model_outputs.densities[q];
+
+          const unsigned int porosity_index = introspection.compositional_index_for_name("porosity");
+          const double porosity = std::max(scratch.material_model_inputs.composition[q][porosity_index],0.000);
+          bulk_density = (1.0 - porosity) * scratch.material_model_outputs.densities[q] + porosity * melt_outputs->fluid_densities[q];
```

Comment:
You set the bulk density above, and overwrite it here. Just declare it `const` here.